File: Data Structures & Algorithms/subsets/submission-1.cpp
class Solution {
public:
    void call(vector<int>& nums , int i, vector<int>& temp ,vector<vector<int>>&ans){
        if(i== nums.size()){
            ans.push_back(temp) ;
            return ;
        }
        temp.push_back(nums[i]) ; 
        call(nums , i+1 , temp , ans) ;
        temp.pop_back() ;
        call(nums, i+1 , temp  , ans);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        int n = nums.size() ; 
        vector<vector<int>> ans ; 
        vector<int> temp ; 
        call(nums , 0 , temp , ans) ; 
        return ans ;
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-1.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        // Arrays.sort(nums);
        List<List<Integer>>ans=new ArrayList<>();
        List<Integer> list=new ArrayList<>();
        call(ans,list,nums,0,target);
        return ans;
    }
    public void call(List<List<Integer>>ans,List<Integer>list,int[]nums,int index,int target){
        if(target<0){
            return;
        }
        if(target==0){
            ans.add(List.copyOf(list));
            return ;
        }
        for(int i=index;i<nums.length;i++){
            list.add(nums[i]);
            call(ans,list,nums,i,target-nums[i]);
            list.remove(list.size()-1);
        }
    }
}